### PR TITLE
Editor + code-boxes v.0.0.1

### PR DESCRIPTION
**Pull request by humanamburu**
Initial implementation of code-boxes and editor.

### CODE REVIEW

File: app/common/components/code/code.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import Beautifier from 'js-beautify';
+
+import 'codemirror/lib/codemirror.css';
+import 'common/styles/code-theme.styl';
+import './code.styl';
+
+//imports specific language mode for editor
+import 'codemirror/mode/javascript/javascript';
+
+export default class Code extends Component {
```
Comment (kirilknysh) [inline app/common/components/code/code.js]: CodeBlock? CodeBox? Code*, please

File: app/game/components/editor/editor.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import 'codemirror/addon/hint/javascript-hint';
+import 'codemirror/addon/hint/show-hint';
+
+import 'codemirror/lib/codemirror.css';
+import 'codemirror/addon/hint/show-hint.css';
+import 'common/styles/code-theme.styl';
+import './editor.styl';
+
+//TODO: Fix autocomplete issues
+export default class Editor extends Component {
```
Comment (kirilknysh) [inline app/game/components/editor/editor.js]: Not critical but I'd name it `CodeInput`

Comment (DaQuirm) [reply app/game/components/editor/editor.js]: Or `CodeEditor`?

File: app/game/components/editor/editor.styl [hunk 3]
```diff
@@ -0,0 +1,16 @@
+@import '~common/styles/code-colors.styl'
+
+//TODO: move lines calculations to shared place
+number-of-lines = 10
+font-size = 14px
+code-mirror-height = number-of-lines * font-size
+
+.code-editor
+    flex-basis auto
```
Comment (DaQuirm) [inline app/game/components/editor/editor.styl]: Is this an override?

File: app/game/components/editor/helpers.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const NOT_VALID_SYMBOLS = '(){};:,';
+
+export function isNewLineEvent(event) {
+    const { origin, removed, text } = event;
+
+    return (origin === '+input' || origin === '+delete') && removed[0] === '' && text[0] === '';
```
Comment (DaQuirm) [inline app/game/components/editor/helpers.js]: `removed` and `text` can be destructured further in the pattern above

File: app/game/components/editor/editor.js (hunk 2 above)
Comment (humanamburu) [reply app/game/components/editor/editor.js]: Was renamed to `CodeInput`

File: app/game/components/editor/editor.styl (hunk 3 above)
Comment (humanamburu) [reply app/game/components/editor/editor.styl]: Oh, thanks, it is realy redundant

File: app/game/components/editor/helpers.js (hunk 4 above)
Comment (humanamburu) [reply app/game/components/editor/helpers.js]: I guess, that nested destructuring has poor readability

File: app/game/components/code-input/code-input.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    componentDidUpdate() {
+        if (this.value !== this.props.value) {
+            this.value = this.props.value;
+        }
+    }
+
+    componentWillUnmount() {
+        if (this.props.enableAutocomplete) {
+            this.codeInput.off('change', this.onChange.bind(this));
```
Comment (Shastel) [inline app/game/components/code-input/code-input.js]: probably you also should assign `null` to  `this.codeInput`, shouldn't you?

**Comment by humanamburu** [comment]
Update:

**Review by kirilknysh — APPROVED** [review]

**Review by Shastel — APPROVED** [review]

**Pull request merged**